src/Features/Core/Portable/SyncNamespaces/AbstractSyncNamespacesSevice.cs

```diff
@@ -0,0 +1,171 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Analyzers.MatchFolderAndNamespace;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CodeFixes.MatchFolderAndNamespace;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.SyncNamespaces
+{
+    internal abstract class AbstractSyncNamespacesSevice<TSyntaxKind, TNamespaceSyntax>
+        : ISyncNamespacesService
+        where TSyntaxKind : struct
+        where TNamespaceSyntax : SyntaxNode
+    {
+        public abstract AbstractMatchFolderAndNamespaceDiagnosticAnalyzer<TSyntaxKind, TNamespaceSyntax> DiagnosticAnalyzer { get; }
+        public abstract AbstractChangeNamespaceToMatchFolderCodeFixProvider CodeFixProvider { get; }
+
+        public async Task<Solution> SyncNamespacesAsync(
+            ImmutableArray<Project> projects,
+            CancellationToken cancellationToken)
+        {
+            var solution = projects[0].Solution;
+            var diagnosticAnalyzers = ImmutableArray.Create<DiagnosticAnalyzer>(DiagnosticAnalyzer);
+            var diagnosticsByProject = await GetDiagnosticsByProjectAsync(projects, diagnosticAnalyzers, cancellationToken).ConfigureAwait(false);
+
+            // If no diagnostics are reported, then there is nothing to fix.
+            if (diagnosticsByProject.Values.All(diagnostics => diagnostics.IsEmpty))
+            {
+                return solution;
+            }
+
+            var fixAllContext = await GetFixAllContextAsync(solution, CodeFixProvider, diagnosticsByProject, cancellationToken).ConfigureAwait(false);
+            var fixAllProvider = CodeFixProvider.GetFixAllProvider();
+            RoslynDebug.AssertNotNull(fixAllProvider);
+
+            return await ApplyCodeFixAsync(fixAllProvider, fixAllContext, cancellationToken).ConfigureAwait(false);
+        }
+
+        private static async Task<ImmutableDictionary<Project, ImmutableArray<Diagnostic>>> GetDiagnosticsByProjectAsync(
```

Comment:
```suggestion
        private static async Task<ImmutableArray<(Project project, ImmutableArray<Diagnostic> diagnostics)>> GetDiagnosticsByProjectAsync(
```

This is a little more verbose in the signature but avoids hashing.

Reply:
But I rely on hashing.